source/unified-test-format/unified-test-format.rst

```diff
@@ -2492,7 +2560,224 @@ An example of this operation follows::
+
+The following arguments are supported:
+
+- ``thread``: Required string. Thread entity that should be stopped and awaited
+  for completion.
+
+An example of this operation follows::
+
+    - name: waitForThread
+      object: testRunner
+      arguments:
+        thread: *thread0
+  
+
+waitForEvent
+~~~~~~~~~~~~
+
+The "waitForEvent" operation instructs the test runner to wait until the
+specified MongoClient has published a specific, matching event a given number of
+times. Note that this includes any events published before the waitForEvent
+operation started.
+
+The following arguments are suported:
+
+- ``client``: Required string. Client entity whose events the runner will be
+  waiting for.
+
+- ``event``: Required `expectedEvent`_ object. The event which the test runner
+  is waiting to be observed on the provided client. The assertions laid out in
+  `expectedEvent`_ MUST be used to determine if an observed event matches
+  ``event``. ``event`` SHOULD have an event type that was included in the
+  ``client``'s ``observeEvents`` field.
+
+- ``count``: Required integer. The number of matching events to wait for before
+  resuming execution of subsequent operations.
+
+For example, the following instructs the test runner to wait for at least one
+poolClearedEvent to be published::
+
+    - name: waitForEvent
+      object: testRunner
+      arguments:
+        client: *client0
+        event:
+          poolClearedEvent: {}
+        count: 1
+
+
+assertEventCount
+~~~~~~~~~~~~~~~~
+
+The "assertEventCount" operation instructs the test runner to assert the
+specified MongoClient has published a specific, matching event a given number of
+times so far in the test. Note that this includes any events published before
+the waitForEvent operation started.
 
+The following arguments are supported:
+
+- ``client``: Required string. Client entity whose events the runner will be
+  counting.
+
+- ``event``: Required `expectedEvent`_ object. The event which the test runner
+  will be counting on the provided client. The assertions laid out in
+  `expectedEvent`_ MUST be used to determine if an observed event matches
+  ``event``. ``event`` SHOULD have an event type that was included in the
+  ``client``'s ``observeEvents`` field.
+
+- ``count``: Required integer. The exact number of matching events that
+  ``client`` should have seen so far.
+
+For example, the following instructs the test runner to assert that
+a single PoolClearedEvent was published::
+
+      - name: assertEventCount
+        object: testRunner
+        arguments:
+          client: *client0
+          event:
+            poolClearedEvent: {}
+          count: 1
+
+recordTopologyDescription
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The "recordTopologyDescription" operation instructs the test runner to retrieve
+the specified MongoClient's current `TopologyDescription <entity_topologydescription_>`_ and store it in
+the `Entity Map`_.
```

Comment:
Drivers aren't required to have a public API to access the current TopologyDescription, are they? I likely won't run these tests in PHPLIB anyway (we defer to libmongoc for SDAM), but the only way I could imagine doing this there would be to configure an SDAM event subscriber behind the scenes and just save the most recent TopologyDescription from the last TopologyDescriptionChanged event we observe.

Reply:
So Rust implemented this via a test-only getter which isn't exported in the regular public API. Swift doesn't implement these tests, but it could have a similar mechanism. I'm not sure if other drivers can do this, but based on the currently existing "recordPrimary" and "waitForPrimaryChange" operations in the old integration tests, I assume that all drivers are capable of something like this. Your proposed implementation certainly sounds like it would also work.